Add set forecast unit

nikosbosse opened the pull request:
Add a helper function to manually set the forecast unit. Addresses #268.

seabbs commented on DESCRIPTION:
```diff
@@ -1,6 +1,6 @@
 Package: scoringutils
 Title: Utilities for Scoring and Assessing Predictions
-Version: 1.1.4
+Version: 1.1.5
```
Do you want to identify that this is a development release or is this a release as is?

seabbs replied:
So the end number here is the minor release number. EpiNow2 has a dev number. If you use that approach then for example pkgdown will build a dev version or a released version of the package. It is also clear when a release (in this case 1.1.5 is done or if its still underwork). At the moment here people could install the same release but get different behaviour. I don't feel that strongly here so really just an FYI.

seabbs replied:
@nikosbosse resolve as you wish

nikosbosse replied:
Just had a look at this: https://combine-australia.github.io/r-pkg-dev/versioning.html
There is a second PR coming once that one is merged to add some documentation, I suggest keeping this for now and making it a minor release in the second PR.

seabbs commented on NEWS.md:
```diff
@@ -1,3 +1,10 @@
+# scoringutils 1.1.5
+
+## Feature updates
+- Added a new function, `set_forecast_unit()` that allows the user to set the 
```
why might they want to do this?

seabbs commented on R/convenience-functions.R:
```diff
@@ -209,3 +209,68 @@ log_shift <- function(x,
   }
   log(x + offset, base = base)
 }
+
```
lots of random spaces

seabbs commented on R/convenience-functions.R:
```diff
@@ -209,3 +209,68 @@ log_shift <- function(x,
+
+#' @title Set unit of a single forecast manually
+#'
+#' @description Helper function to set the unit of a single forecast manually.
```
Link here to the idea that otherwise will happen automatically and touch on why users might not want this to be the case (and hence why they might use this function)

seabbs commented on R/convenience-functions.R:
```diff
@@ -209,3 +209,68 @@ log_shift <- function(x,
+#'
+#' @inheritParams score
+#' @param forecast_unit character vector with the names of the columns that
+#' uniquely identify a single forecast
```
full stop and caps for sentence start

seabbs commented on R/convenience-functions.R:
```diff
@@ -209,3 +209,68 @@ log_shift <- function(x,
+#' @importFrom data.table ':=' is.data.table copy
+#' @author Nikos Bosse \email{nikosbosse@@gmail.com}
+#' @export
+#' @references Transformation of forecasts for evaluating predictive
```
why is this reference here?

seabbs replied:
?

nikosbosse replied:
Aaah those conversations were hidden!
You can't ever link enough to the paper! :D 
Probably copypasting from somewhere.. I removed it :)

seabbs commented on R/convenience-functions.R:
```diff
@@ -209,3 +209,68 @@ log_shift <- function(x,
+#' medRxiv 2023.01.23.23284722
+#' \doi{https://doi.org/10.1101/2023.01.23.23284722}
+#' <https://www.medrxiv.org/content/10.1101/2023.01.23.23284722v1> # nolint
+
```
random space

seabbs commented on R/convenience-functions.R:
```diff
@@ -209,3 +209,68 @@ log_shift <- function(x,
+#'   example_quantile,
+#'   c("location", "target_end_date", "target_type", "horizon", "model")
+#' )
+#'
```
random empty line

seabbs commented on R/convenience-functions.R:
```diff
@@ -209,3 +209,68 @@ log_shift <- function(x,
+
+  if (length(missing) > 0) {
+    warning(
+      paste0(
```
you don't need to use paste0 here - can use warning directly.

nikosbosse replied:
oh... I wasn't aware!

seabbs commented on R/convenience-functions.R:
```diff
@@ -209,3 +209,68 @@ log_shift <- function(x,
+#'
+
+set_forecast_unit <- function(data,
+                              forecast_unit = NULL) {
```
is this really over 80L and so needs a linebreak?

seabbs commented on R/convenience-functions.R:
```diff
@@ -209,3 +209,68 @@ log_shift <- function(x,
+    )
+  }
+
+  keep_cols <- intersect(
```
Should it warn if target columns are not present?

nikosbosse replied:
the lines above should produce a warning. But indeed I forgot to remove the non-existing columns from the vector.

seabbs commented on R/utils.R:
```diff
@@ -256,3 +251,26 @@ get_forecast_unit <- function(data, prediction_type) {
   forecast_unit <- setdiff(colnames(data), protected_columns)
   return(forecast_unit)
 }
+
```
random spaces

seabbs reviewed (commented):
Overall LGTM to me but a few niggles and automated lining issues. I also imagine you want to talk about this function/option in the docs for score or check_forecasts and potentially provide an example in the news file. 

This is based on an issue right? It should be linked to the PR and in the news item for this change. 

It would also be nice to hear what is happening with the 5% coverage drop for convenience functions and why we think this is okay.

nikosbosse commented:
I think I addressed all comments so from my side this could be merged. I have a separate PR that finally makes `check_forecast()` pipeable and updates the documentation.

seabbs commented on NEWS.md:
```diff
@@ -1,3 +1,18 @@
+- Added a new function, `set_forecast_unit()` that allows the user to set the 
+forecast unit manually. The function removes all columns that are not relevant
+for uniquely identifying a single forecast. If not done manually, `scoringutils` 
+attempts to determine the unit
```
some weird linebreak stuff showing here for me?

nikosbosse replied:
weird.. I removed some linebreaks, so should be fixed hopefully.

seabbs commented on NEWS.md:
```diff
@@ -1,3 +1,18 @@
+behaviour, so setting the forecast unit explicitly can help make the code
+easier to debug and easier to read (see issue #268). 
+When used as part of a workflow,
+`set_forecast_unit()` can then directly be piped into `check_forecasts()` to
```
can be vs can then

nikosbosse replied:
removed "then" assuming that's what you meant

seabbs commented on R/convenience-functions.R:
```diff
@@ -209,3 +202,61 @@ log_shift <- function(x,
+#' additional protected columns, e.g. for true values, predictions or quantile
+#' levels) and removes duplicate rows.
+#' If not done manually, `scoringutils` attempts to determine the unit
+#' of a single automatically by simply assuming that all column names are
```
a single forecast

nikosbosse replied:
good spot, fixed

seabbs commented on R/convenience-functions.R:
```diff
@@ -209,3 +202,61 @@ log_shift <- function(x,
+#' levels) and removes duplicate rows.
+#' If not done manually, `scoringutils` attempts to determine the unit
+#' of a single automatically by simply assuming that all column names are
+#' relevant to determine the forecast unit. This can lead to unexpected
```
may lead

nikosbosse replied:
Changed "can lead" to "may lead"

seabbs commented on R/convenience-functions.R:
```diff
@@ -209,3 +202,61 @@ log_shift <- function(x,
+#' relevant to determine the forecast unit. This can lead to unexpected
+#' behaviour, so setting the forecast unit explicitly can help make the code
+#' easier to debug and easier to read. When used as part of a workflow,
+#' `set_forecast_unit()` can then directly be piped into `check_forecasts()` to
```
can be vs can then

nikosbosse replied:
fixed

seabbs commented on R/convenience-functions.R:
```diff
@@ -209,3 +202,61 @@ log_shift <- function(x,
+set_forecast_unit <- function(data, forecast_unit = NULL) {
+
+  datacols <- colnames(data)
+  missing <- forecast_unit[!(forecast_unit %in% datacols)]
```
what does using the default (NULL) do exactly? Does it just end up doing the internal score behaviour of using all columns? If yes should there not be some verbose messaging to this effect?

seabbs replied:
If this is how it behaves could it be used inside check_forecasts when the unit is determined?

nikosbosse replied:
You're right, having a default NULL doesn't really make sense. As is it would return a data.frame with only the 'protected columns' (i.e. thinks like sample or quantile or true_value). I removed the default. 
`check_forecasts()` uses a function `get_forecast_unit()` instead.

seabbs reviewed (commented):
LGTM to me in the main but there are still some outstanding conversations and I think we haven't discussed the reduction in test coverage? 

Address as you wish and give me a review ping.

nikosbosse commented:
Thanks a lot! Overlooked a hidden comment. 
The coverage drop should have been addressed, I had added another test.. Do you still see a drop? Need to recheck then. 
Otherwise should be good for another review!

seabbs reviewed (approved):
All looks good to me now. Rerunnning code cov shows you covered the test gap - nice job! 

🥳

nikosbosse commented:
Thanks a lot! :)
